sdk/iot/docs/iot_hub_flow.puml

```diff
@@ -0,0 +1,146 @@
+@startuml

```

Comment:
actually one comment on this docs directory. Would it be possible to put all the "resources" in their own directory and have the top level of `docs/` be only full fledged documentation? So only the readme's in the top level and everything else like pictures, raw `puml` files, etc in `docs/resources`. I fear that as we add more documentation users will have to comb through a lot of "junk" they don't care about before getting to the docs they do care about.

Reply:
Makes sense: as you probably noticed, there are other things there that shouldn't, like samples.

I propose:

```
\readme.md
\docs
    \resources
\samples
    \paho
\inc (flat)
\src (flat)
\test
```

@ewertons is this in-line with what you're proposing? I'd like to keep things alligned in this PR to make it easier for me to rebase after you merge.

Reply:
I like it. The top top level readme explaining the iot directory should go where it is as you just commented. Any supplemental documentation should go in `docs/` and any resources that are used to create/add graphics etc to documentation should go in `docs/resources`

Reply:
I'm not opposed to this structure.  However it would be the only service with a service level docs folder.
https://azure.github.io/azure-sdk/policies_repostructure.html

Is there a reason for not placing all the documentation in the iot/README.md and not placing the samples in the samples folder?
  /iot/samples/....

Reply:
@gilbertw apologizes, I didn't know that spec existed.

> Is there a reason for not placing all the documentation in the iot/README.md and not placing the samples in the samples folder?

Won't fit: we will have one design doc for each service feature (telemetry, c2d, methods, twin, etc). The one I have posted here is the overall state machine and how-to guide.

From the spec: I don't understand the reason for doc\devdoc to be in root. Is that the place to put our design docs required for devs wanting to contribute? Why is this repo not having one?

> Is there a reason for not placing all the documentation in the iot/README.md and not placing the samples in the samples folder?

Samples will go in /iot/samples. Each app in it's own folder. We will have ports for various MQTT clients as well as platforms (why I plan to have a paho folder).
I will modify the current PR. This was sent out mainly to start the API design & review discussion.